File: okta-skill/okta-ciam-developer/scripts/04-m2m-client-credentials/python/client_credentials.py
# Client Credentials with private-key JWT (recommended client auth).
# Source pattern: developer.okta.com/docs/guides/implement-grant-type/clientcreds/main/
# pip install requests pyjwt cryptography

import os
import time
import uuid
import jwt
import requests

ISSUER = os.environ["OKTA_OAUTH2_ISSUER"]
CLIENT_ID = os.environ["OKTA_CLIENT_ID"]
PRIVATE_KEY = os.environ["OKTA_PRIVATE_KEY"]  # PEM
KEY_ID = os.environ.get("OKTA_KEY_ID")        # kid registered in the app JWKS

_cache = {"token": None, "exp": 0}


def _client_assertion() -> str:
    now = int(time.time())
    return jwt.encode(
        {
            "iss": CLIENT_ID,
            "sub": CLIENT_ID,
            "aud": f"{ISSUER}/v1/token",
            "iat": now,
            "exp": now + 300,
            "jti": str(uuid.uuid4()),
        },
        PRIVATE_KEY,
        algorithm="RS256",
        headers={"kid": KEY_ID} if KEY_ID else None,
    )


def get_token(scope: str = "orders:read") -> str:
    if _cache["token"] and _cache["exp"] > time.time() + 30:
        return _cache["token"]

    resp = requests.post(
        f"{ISSUER}/v1/token",
        data={
            "grant_type": "client_credentials",
            "scope": scope,
            "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
            "client_assertion": _client_assertion(),
        },
        timeout=10,
    )
    resp.raise_for_status()
    body = resp.json()
    _cache.update(token=body["access_token"], exp=time.time() + body["expires_in"])
    return _cache["token"]


if __name__ == "__main__":
    token = get_token()
    print("Got token:", token[:40], "...")
    # requests.get("https://api.example.com/orders",
    #              headers={"Authorization": f"Bearer {token}"})

File: okta-skill/okta-ciam-developer-copilot/skills/okta-ciam-developer/scripts/04-m2m-client-credentials/node/client-credentials.js
// Client Credentials flow — both client auth methods.
// Source pattern: developer.okta.com/docs/guides/implement-grant-type/clientcreds/main/
// npm i jsonwebtoken (only for the private-key JWT variant)

const jwt = require('jsonwebtoken');

const ISSUER = process.env.OKTA_OAUTH2_ISSUER; // custom server for your API scopes;
// use https://{org}.okta.com/oauth2 (org server) for okta.* management scopes.
const CLIENT_ID = process.env.OKTA_CLIENT_ID;

// --- Variant A: client secret (simpler, weaker) -----------------------------
async function tokenWithSecret(scope) {
  const res = await fetch(`${ISSUER}/v1/token`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
      Authorization:
        'Basic ' + Buffer.from(`${CLIENT_ID}:${process.env.OKTA_CLIENT_SECRET}`).toString('base64'),
    },
    body: new URLSearchParams({ grant_type: 'client_credentials', scope }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json(); // { access_token, expires_in, scope, token_type }
}

// --- Variant B: private-key JWT (recommended) --------------------------------
function buildClientAssertion(privateKeyPem) {
  const now = Math.floor(Date.now() / 1000);
  return jwt.sign(
    {
      iss: CLIENT_ID,
      sub: CLIENT_ID,
      aud: `${ISSUER}/v1/token`,
      iat: now,
      exp: now + 300,
      jti: crypto.randomUUID(),
    },
    privateKeyPem,
    { algorithm: 'RS256', keyid: process.env.OKTA_KEY_ID } // kid must match registered JWKS
  );
}

async function tokenWithPrivateKey(scope) {
  const assertion = buildClientAssertion(process.env.OKTA_PRIVATE_KEY);
  const res = await fetch(`${ISSUER}/v1/token`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      grant_type: 'client_credentials',
      scope,
      client_assertion_type: 'urn:ietf:params:oauth:client-assertion-type:jwt-bearer',
      client_assertion: assertion,
    }),
  });
  if (!res.ok) throw new Error(await res.text());
  return res.json();
}

// Simple in-memory token cache — always cache until expiry
let cached = null;
async function getToken(scope = 'orders:read') {
  if (cached && cached.expiresAt > Date.now() + 30_000) return cached.token;
  const t = await tokenWithPrivateKey(scope);
  cached = { token: t.access_token, expiresAt: Date.now() + t.expires_in * 1000 };
  return cached.token;
}

module.exports = { getToken, tokenWithSecret, tokenWithPrivateKey };
